### DISCUSSION TITLE
Feature request: Api wrapper class

### DISCUSSION THREAD

**Discussion by Radiergummi** [craftcms/element-api#146] [Ideas]
Hey there, I'm quite new to Craft, so if this won't fit in the conventions, be free to close this issue.  
Coming mainly from Laravel, the syntax to define endpoints seems a little off to me. I'd favor to use something like `$api->addEndpoint(string $uri, Closure $handler);`. On the one hand, that'd make IDE hints a tad more useful. On the other, it'd be nice to have a `$api->addCollection(string $uri, Closure $endpointsCallback);`, too. Within this `$endpointsCallback`, one could define additional endpoints sharing the collection URI as a prefix. That'd allow for conveniently splitting the `elementapi.php` into multiple files once the number of endpoints grows.  
I've made a little proof of concept class which you'll find in [in this gist](https://gist.github.com/Radiergummi/3c402345f249a22b75a8afa240a1fba5), including usage examples. This could obviously be extended in functionality.  

What's your opinion on this?

**Comment by brandonkelly** [comment]
Thanks for the suggestion!

That’s certainly an elegant way of defining endpoints so I’m not against it, but it’s not really feasible right now, because plugin settings files [get loaded](https://github.com/craftcms/cms/blob/develop/src/services/Plugins.php#L580) before the plugin has even been instantiated.

I’ll leave the issue open though, as in the long run it would be nice if we could come up with a way to make it possible.
